Sources/Parsing/Parsers/CaseIterableRawRepresentable.swift

```diff
@@ -0,0 +1,133 @@
+extension CaseIterable where Self: RawRepresentable, RawValue == Int {
```

Comment:
@mbrandonw would `RawValue: LosslessStringConvertible & Comparable` also work here?

Reply:
@iampatbrown If I'm not mistaken, the restriction can theoretically be narrowed to being expressible by an integer or floating-point literal, because that's ultimately what will be allowed for `RawRepresentable` enums (other than those with `RawValue` being `ExpressibleByStringLiteral)`. So we can cover every situation with 3 overloads (`ExpressibleBy(Integer/Float/String)Literal`), but I don't know if we can easily extract parser/printers from it.

Being `Comparable` allows to reach `UInt`s for example, but it doesn't help with custom made types. For example, this compiles:
```swift
struct Value: RawRepresentable, Equatable {
  var rawValue: Int
  init(rawValue: Int) { self.rawValue = rawValue }
}

extension Value: ExpressibleByIntegerLiteral {
  init(integerLiteral value: Int) {
    self.rawValue = value
  }
}

enum Enum: Value, CaseIterable {
  case first = 1
}
```

`String` and `Int` are already covering the vast majority of the cases (!) though, but I 'm afraid that `LosslessStringConvertible & Compable` will only partially improve the situation (the `Enum` above would pass through).

Reply:
@tgrapperon if I'm not mistaken, the reason `RawValue == Int` works is because `Int` conforms to `LosslessStringConvertible & Comparable`.  The `String.init` being used in `toPrefix: { String($0) }` is `init<T>(_ value: T) where T : LosslessStringConvertible`. 

`RawValue: LosslessStringConvertible & Comparable` should work for `FixedWidthInteger`, `Double` and `Substring`. As for printing, I'm pretty sure that would remain the same using `toPrefix`:

```swift
input.prepend(contentsOf: self.toPrefix(output.rawValue))
```

Maybe I'm missing something though :P

Reply:
@iampatbrown You're right about `LosslessStringConvertible`. I was only reasoning at the `RawRepresentable` enum level, without considering how we can extract parsers/printers from the `RawValue` later.

You're probably right for the constraint, and the example I've provided is quite exotic and a dedicated parser can always be implemented if it ever happens in the wild. Furthermore, the new `RawValue` types that can be reached are all comparable, so it directly works with respect to the comment I made about sorting below.

I didn't check if the compiler was working the same with `Int`/`String` or `LosslessStringConvertible&Comparable`, but if it is, it looks like a better alternative.